synapse/lib/coro.py

```diff
@@ -2,102 +2,211 @@
 Async/Coroutine related utilities.
 '''
 import asyncio
+import inspect
 import logging
+import threading
+import collections
 
 logger = logging.getLogger(__name__)
 
 import synapse.glob as s_glob
+import synapse.lib.base as s_base
 
-class Fini:
+def iscoro(item):
+    return inspect.iscoroutine(item)
 
-    def __init__(self):
-        self.finis = []
-        self.isfini = False
-        self.entered = False
-        self.exitinfo = None
+class Queue(s_base.Base):
```

Comment:
We should probably consider rolling all of this up into `synapse.lib.queue` at some point. It feels like a code smell to have Queue style objects defined in three different files.